### DISCUSSION TITLE
How to import a python array into the emitter

### DISCUSSION THREAD

**Discussion by yangocean-sudo** [mitsuba-renderer/mitsuba3#677] [Q&A] [answered]
Hi, in the last [discussion](https://github.com/mitsuba-renderer/mitsuba3/discussions/614), we successfully imported the binary file into the emitter and using `fs::path file_path = fs->resolve(props.string("filename"));` to read the file. Now, we want to generate the array of photons in Python and directly import it into the emitter in the scene we created in Python, like:
'type': 'photon_emitter',
'photon_list': [0.0, 1.0, 0.0, 1.0,2.0,3.0, ... ]
I tried `props.get<std::vector<float>>("photon_list");` or `props.get<Float>("photon_list");` in the emitter cpp file, but none of them works. It seems there is no direct way for the property to read a float array. Can you give me some suggestions? Thank you for your time.

**Comment by njroussel** [comment, marked as answer]
Hi @yangocean-sudo 

You are correct, this is not supported.

I would suggest looking at the `bitmap` plugin or the `grid` plugin. Both can accept a runtime constructed object in their constructor `props`.
For example the `grid` ([documentation](https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_volumes.html#grid-based-volume-data-source-gridvolume)) will accept a `VolumeGrid` object that is created at runtime. And a `VolumeGrid` has a special [Python constructor](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/render/python/volumegrid_v.cpp#L9-L39) that accepts NumPy types. The idea would be to create something similar to a `VolumeGrid` object. (If you're feeling lazy you can most likely re-use the `VolumeGrid` object and just treat it as a buffer)

**Reply by yangocean-sudo** [reply]
It's very helpful. Thanks a lot.
